## Workflow Overview

### Fusion Transcript Detection: Short Read RNA-Seq Data

1. Run FusionCatcher Analysis (fusioncatcher_run.sh):  
   - Processes a list of directories containing FASTQ files.  
   - Executes the FusionCatcher program to detect fusion genes.  
   - Saves the results in corresponding output directories.  

2. Align Reads to Fusion Transcript Junction Sequences (bowtie2_alignment_mayo.py and bowtie2_alignment_hipsci.py):  
   - Takes fusion transcripts identified by FusionCatcher and maps supporting reads to the junction sequences as a custom reference using Bowtie2.  
   - This step uses two different scripts to accommodate the naming conventions of specific datasets:  
     - bowtie2_alignment_mayo.py: Designed for the Mayo RNA sequencing study.  
     - bowtie2_alignment_hipsci.py: Designed for the HipSci RNA sequencing dataset.  
   - Steps performed by these scripts:
     - Align reads to indexed fusion transcripts using Bowtie2.  
     - Convert SAM to BAM files, sort, and index them.  
     - Extract supporting reads for each fusion transcript.  
   - Outputs:
     - SAM/BAM files stored in structured directories.  
     - FASTA files for each fusion transcript and sample combination.  

3. Count Reads Mapped to Fusion Transcripts (bowtie2_counts.py):  
   - Counts the number of reads mapped to each fusion transcript for each sample using the FASTA files generated from Bowtie2 alignments.  
   - Outputs results as Counts_pivot_[date+time].txt.  
   - Results are used as input for FusionCatcher_processing.R and variable_fusion_expression.R.

---

### Fusion Transcript Detection: Long Read RNA-Seq Data

4. Construct Transcript Models (isoquant.py):  
   - Uses FASTA files containing PacBio reads supporting fusion transcripts to construct transcript models.  
   - Saves results in subdirectories corresponding to the input file names.

---

### Fusion Peptide Detection: Ribosome Profiling Data

5. Generate Random Sets of Control Transcript Exon Junctions
   - Retrieves exon and intron coordinates of exon junctions of all genes from GENCODE v44 (final_posctrl_creation_0.sh and final_posctrl_creation_python_to_run_with_0.py)
   - Randomly selects same number control exon junctions as fusion transcript set (final_posctrl_creation_1.sh)
   - Maps ENSG names to ENST names (final_posctrl_creation_2.py)
   - Creates FASTA outputs with final control exon junction sequences (final_posctrl_creation_3.sh and final_posctrl_creation_4.py)

6. Align Ribo-Seq Reads to Fusion Transcript or Control Transcript Junction Sequences (Ribo-seq_FT.py):
   - Processes fastq file containing Ribo-Seq reads.
   - Aligns reads to indexed transcripts using Bowtie2.
   - Converts SAM to BAM files, sorts, and indexes the BAM files.
   - Extracts reads for each transcript.
   - Saves results in structured directories:
        - SAM/BAM files for each fusion transcript and sample combination.
        - FASTA files with headers assigned based on the gene name the RS reads matched.
   - Filters out Ribo-Seq reads that do not span the fusion junction.

7. Filter control transcript mapped reads with with BlastN (Ribo-seq_pos_ctrl_BlastN.py):
   - Uses BlastN to identify and filter out Ribo-Seq reads that map to more than one Ensembl gene.

---

### Genotyping: WGS Data

8. Perform Coverage Analysis for WGS BAM Files (coverage_analysis.sh):  
   - Analyzes coverage in specified regions of HipSci WGS BAM files.  
   - This step is required to cross-reference genomic copy number in specific regions with fusion transcript expression.  
   - Requires the following files:  
     - CSV file (hipsci-files_wgs_cram.csv) with metadata for cell lines (file provided here; obtained from hipsci.org).  
     - Optional text file (cell_lines.txt) listing specific cell lines to analyze.  
     - BED file (copy_regions.bed) specifying regions of interest.  
   - Results are used as input for variable_fusion_expression.R.

9. Calculate Read Count Ratios (calculate_ratios.py):  
   - Combines read counts from multiple VCF files generated by coverage_analysis.sh.  
   - Calculates ratios of read counts for specific regions and groups the results by region type.  
   - Results are used as input for variable_fusion_expression.R.

10. Perform SNP Genotyping (H1H2_SNP_genotyping.py):  
   - Performs SNP genotyping for H1 and H2 cell lines using SAMtools and BCFtools.  
   - Requires BAM files generated from coverage_analysis.sh.  
   - Results are used as input for variable_fusion_expression.R.

---

### Data Processing and Visualization

11. Visualize fusion transcript expression (FusionCatcher_processing.R):  
   - Takes outputs from short-read RNA-Seq analysis and combines them for visualization.  

12. Analyze Expression of Fusion Transcripts with Inter-Individual Variation (variable_fusion_expression.R):  
   - Focuses on fusion transcripts with inter-individual variation.  
   - Combines outputs from short-read RNA-Seq and WGS for targeted visualization.

13. Visualize Coding Potential of Fusion Transcript Dataset (coding_potential_bubblechart.py and coding_potential_piechart.py)
   - Categorizes protein-coding potential of fusion transcripts based on upstream and downstream fusion junction locations.
 
14. Visualize Number of Control and Fusion Transcripts with Supporting Ribo-seq Reads (Ribo-seq_funnelchart.py)
   - Illustrates numbers of fusion transcripts and randomly selected control transcripts detected in Ribo-seq data.

---

## Repository Contents

### Fusion Transcript Detection: Short Read RNA-Seq Data
- fusioncatcher_run.sh  
- bowtie2_alignment_mayo.py  
- bowtie2_alignment_hipsci.py  
- bowtie2_counts.py  

### Fusion Transcript Detection: Long Read RNA-Seq Data
- isoquant.py  

### Fusion Peptide Detection: Ribosome Profiling Data
- Ribo-seq_FT.py
- Ribo-seq_pos_ctrl_BlastN.py
- Ribo-seq_positive_control_generation:
   - final_posctrl_creation_0.sh
   - final_posctrl_creation_python_to_run_with_0.py
   - final_posctrl_creation_1.sh
   - final_posctrl_creation_2.py
   - final_posctrl_creation_3.sh
   - final_posctrl_creation_4.py

### Genotyping: WGS Data
- coverage_analysis.sh  
- calculate_ratios.py  
- H1H2_SNP_genotyping.py  

### Data Processing and Visualization
- FusionCatcher_processing.R  
- variable_fusion_expression.R
- coding_potential_bubblechart.py
- coding_potential_piechart.py
- Ribo-seq_funnelchart.py

---

## Dependencies

### Fusion Transcript Detection: Short Read RNA-Seq Data
#### `fusioncatcher_run.sh`
- FusionCatcher (v1.33)  
- conda (v24.3.0)  
- GNU Parallel (v20171222)  
- Bash (v4.4.20(1)-release)  
- SLURM Workload Manager (v23.11.10)  

#### `bowtie2_alignment_mayo.py` and `bowtie2_alignment_hipsci.py`
- Python (v3.8.19) *Note: Requires Python 3.8.19 for compatibility.*  
- Bowtie2 (v2.4.2)  
- Samtools (v1.17)  
- Pysam (built-in Python)
- Biopython (built-in Python)
- Multiprocessing (built-in Python)  
- gzip (built-in Python)  
- Bash (v4.4.20(1)-release)  
- SLURM Workload Manager (v23.11.10)  

#### `bowtie2_counts.py`
- Python (v3.9.17)  
- glob (built-in)  
- datetime (built-in)  
- collections (built-in)  
- Bash (v4.4.20(1)-release)  
- SLURM Workload Manager (v23.11.10)  

---

### Fusion Transcript Detection: Long Read RNA-Seq Data
#### `isoquant.py`
- Python (v3.9.17)  
- IsoQuant (v3.1.2)  
- Multiprocessing (built-in)  
- Subprocess (built-in)  
- Bash (v4.4.20(1)-release)  
- SLURM Workload Manager (v23.11.10)  

---

### Fusion Peptide Detection: Ribosome Profiling Data
#### `Ribo-seq_FT.py`
- Python (v3.9.17)
- conda (v24.3.0)  
- glob (built-in)
- os (built-in)
- pysam (built-in)
- subprocess (built-in)
- blastN (2.15.0+)
- Bash (v4.4.20(1)-release)  
- SLURM Workload Manager (v23.11.10)

#### `Ribo-seq_pos_ctrl_BlastN.py`
- Python (v3.9.17)
- conda (v24.3.0) 
- os (built-in)
- blastN (2.15.0+)

#### `final_posctrl_creation_0.sh`
- Bash (v4.4.20(1)-release)
- Samtools (v1.17)

#### `final_posctrl_creation_python_to_run_with_0.py`
- Python (v3.9.17)
- conda (v24.3.0) 
- pandas (v2.0.2)

#### `final_posctrl_creation_1.sh`
- Bash (v4.4.20(1)-release)
- Samtools (v1.17)  

#### `final_posctrl_creation_2.py`
- Python (v3.9.17)
- conda (v24.3.0) 

#### `final_posctrl_creation_3.sh`
- Bash (v4.4.20(1)-release)  

#### `final_posctrl_creation_4.py`
- Python (v3.9.17)
- conda (v24.3.0) 

---

### Genotyping: WGS Data
#### `coverage_analysis.sh`
- Bash (v4.4.20(1)-release)  
- Samtools (v1.17)  
- Bedtools (v2.31.1)  
- Perl (v5.26.3)  
- grep (built-in v3.1)  
- sed (built-in v4.5)  
- SLURM Workload Manager (v23.11.10)  

#### `calculate_ratios.py`
- Python (v3.9.17)  
- collections (built-in)  
- os (built-in)  
- sys (built-in)  

#### `H1H2_SNP_genotyping.py`
- Python (v3.9.17)  
- Samtools (v1.17)  
- BCFtools (v1.17)  
- pyVCF (v0.6.8)  
- glob (built-in)  
- subprocess (built-in)  
- csv (built-in)  

---

### Data Processing and Visualization
#### `FusionCatcher_processing.R`
- R (v4.3.3)  
- tidyverse (v2.0.0)  
- dplyr (v1.1.4)  
- ggplot2 (v3.5.1)  
- gplots (v3.2.0)  
- tibble (v3.2.1)  
- GenomicRanges (v1.52.1)  
- IRanges (v2.34.1)  
- biomaRt (v2.56.1)  
- karyoploteR (v1.26.0)  
- readxl (v1.4.3)  
- writexl (v1.5.1)  

#### `variable_fusion_expression.R`
- R (v4.3.3)  
- dplyr (v1.1.4)  
- tidyr (v1.3.0)  
- ggplot2 (v3.5.1)  
- stringr (v1.5.1)  
- ggrepel (v0.9.6)  
- car (v3.1-3)  
- FSA (v0.9.5)  

#### `coding_potential_bubblechart.py`
- Python (v3.9.17)
- pandas (v2.0.2)
- matplotlib.pyplot (v3.9.0)

#### `coding_potential_piechart.py`
- Python (v3.9.17)  
- matplotlib.pyplot (v3.9.0)

#### `Ribo-seq_funnelchart.py`
- Python (v3.9.17)
- plotly (v5.23.0)

---

## List of Dependencies (All Scripts)
1. FusionCatcher (v1.33)  
2. conda (v24.3.0)  
3. GNU Parallel (v20171222)  
4. Bowtie2 (v2.4.2)  
5. Samtools (v1.17)  
6. BCFtools (v1.17)  
7. Bedtools (v2.31.1)  
8. IsoQuant (v3.1.2)  
9. pyVCF (v0.6.8)  
10. Perl (v5.26.3)  
11. Bash (v4.4.20(1)-release)  
12. grep (built-in v3.1)  
13. sed (built-in v4.5)  
14. SLURM Workload Manager (v23.11.10)  
15. Python (v3.9.17)  
    - Pysam (built-in)
    - Biopython (built-in)
    - glob (built-in)  
    - datetime (built-in)  
    - collections (built-in)  
    - gzip (built-in)  
    - os (built-in)  
    - sys (built-in)  
    - subprocess (built-in)  
    - multiprocessing (built-in)  
    - csv (built-in)  
    - **Note**: For `bowtie2_alignment` scripts, Python 3.8.19 is required.
    - matplotlib.pyplot (v3.9.0)
    - plotly (v5.23.0)
16. R (v4.3.3)  
    - tidyverse (v2.0.0)  
    - dplyr (v1.1.4)  
    - ggplot2 (v3.5.1)  
    - gplots (v3.2.0)  
    - tibble (v3.2.1)  
    - GenomicRanges (v1.52.1)  
    - IRanges (v2.34.1)  
    - biomaRt (v2.56.1)  
    - karyoploteR (v1.26.0)  
    - readxl (v1.4.3)  
    - writexl (v1.5.1)  
    - stringr (v1.5.1)  
    - ggrepel (v0.9.6)  
    - car (v3.1-3)  
    - FSA (v0.9.5)  

---
